apps/email/js/html_cache_restore.js

```diff
@@ -102,103 +250,236 @@ window.htmlCacheRestorePendingMessage = [];
+    if (ignoreMessage('request-sync')) {
+      return;
+    }
+
+    dispatch('request-sync', [e]);
+  }
+
+  function hydrateHtml(id) {
+    var parsedResults = retrieve(id);
 
     if (parsedResults.langDir) {
       document.querySelector('html').setAttribute('dir', parsedResults.langDir);
     }
 
     var contents = parsedResults.contents;
+
+    // Automatically restore the HTML as soon as module is executed.
+    // ASSUMES card node is available (DOMContentLoaded or execution of
+    // module after DOM node is in doc)
+    var cardsNode = document.getElementById(selfNode.dataset.targetid);
+
     cardsNode.innerHTML = contents;
     startupCacheEventsSent = !!contents;
-    window.addEventListener('load', startApp, false);
+
+    if (contents) {
+      console.log('Using HTML cache for ' + id);
+    }
+
+    if (startupCacheEventsSent) {
+      window.performance.mark('navigationLoaded');
+      window.dispatchEvent(new CustomEvent('moz-chrome-dom-loaded'));
+      window.performance.mark('visuallyLoaded');
+      window.dispatchEvent(new CustomEvent('moz-app-visually-complete'));
+    }
   }
 
-  // START COPY performance_testing_helper.js
-  // A copy instead of a separate script because the gaia build system is not
-  // set up to inline this with our main script element, and we want this work
-  // to be done after the cache restore, but want to trigger the events that
-  // may be met by the cache right away without waiting for another script
-  // load after html_cache_restore.
-  function dispatch(name) {
-    if (!window.mozPerfHasListener) {
+  function finishStartup() {
+    // This can be called multiple times due to pendingMessage listeners
+    // so only do the work once.
+    if (!startingUp) {
       return;
     }
 
-    var now = window.performance.now();
-    var epoch = Date.now();
+    var scriptNode = document.createElement('script');
 
-    setTimeout(function() {
-      var detail = {
-        name: name,
-        timestamp: now,
-        epoch: epoch
-      };
-      var event = new CustomEvent('x-moz-perf', { detail: detail });
+    if (loader) {
+      scriptNode.setAttribute('data-main', loadSrc);
+      scriptNode.src = loader;
+    } else {
+      scriptNode.src = loadSrc;
+    }
 
-      window.dispatchEvent(event);
-    });
+    document.head.appendChild(scriptNode);
+
+    startingUp = false;
   }
 
-  ([
-    'moz-chrome-dom-loaded',
-    'moz-chrome-interactive',
-    'moz-app-visually-complete',
-    'moz-content-interactive',
-    'moz-app-loaded'
-  ].forEach(function(eventName) {
-      window.addEventListener(eventName, function mozPerfLoadHandler() {
-        dispatch(eventName);
-      }, false);
-    }));
+  // mozHasPendingMessage seems like it can only be called once per message
+  // type, so only asking once. These are ordered by least important to most
+  // important. Similar to gateEntry, we only want to process one event on
+  // startup, ignore the others.
```

Comment:
This rationale seems to contradict the rationale in messageTooFast where you suggest 'request-sync' is fine and effectively orthogonal to other-things.  Certainly it does seem undesirable that if the user triggers an activity around the time request-sync wakes us up that we'll ignore the user's actviity.

Specifically, I think:

- It makes sense that only one of notification or activity can win since they both are effectively trying to decide on the state of the UI.  (Arguably we could show the message/inbox under the compose card.  I agree we don't want to do that, though.)

- request-sync should be orthogonal.  In the game of prioritization, I think an argument can be made that request-sync is the least important to deal with promptly, but it should be dealt with.  Given the current pendingMessage flow here, I think the right thing is probably to only set pendingMessage to "request-sync" if none of the prior cases fired.  If they did fire, instead have the triggering of the actual request-sync logic then() on a promise that gets resolved once the UI has reached the user-interactive state.